### PR TITLE
Added basic_NFC

### PR DESCRIPTION
**Pull request by icodk**
This is a sample project that uses Arduino NFC library on a ESP8266 development board like nodemcu or D1 mini. This specific application shows how to read from two RFC antennas like the the one you can get on Aliexpress (search for 13.56MHz RFID Module MF RC522)
Wiring:
```
MISO - D6
MOSI-D7
CLK-D5
SDA-D8
RST- NC
IRQ-NC
```
It can be powered from the +3.3V pin of the nodemcu
Output example:
```
canning nfc Scanner:4 
Card UID on scanner:15:
 06 DF E4 D5
Scanning nfc Scanner:4 
Scanning nfc Scanner:15 
Scanning nfc Scanner:4 
Scanning nfc Scanner:15 
Card UID on scanner:4:
 06 DF E4 D5
Scanning nfc Scanner:15
```

### CODE REVIEW

File: samples/Basic_NFC/Makefile-user.mk [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+## SMING_HOME sets the path where Sming framework is located.
+## Windows:
+# SMING_HOME = c:/tools/sming/Sming 
+
+# MacOS / Linux
+# SMING_HOME = /opt/sming/Sming
+
+## COM port parameter is reqruied to flash firmware correctly.
+## Windows: 
+COM_PORT = COM7
```
Comment (slaff) [inline samples/Basic_NFC/Makefile-user.mk]: The default values for the settings are too specific and will only work on Windows machine (on which the serial adapter is on COM7) and probably and some of the USB serial adapters. Please, replace that file with the Makefile-user.mk from the Basic_Blink example.

File: samples/Basic_NFC/app/application.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+
+
+void init()
+{
+	Serial.begin(SERIAL_BAUD_RATE); // 115200 by default
+
+	procTimer.initializeMs(2000, sayHello).start();
+
+	/// Reading callback example:
```
Comment (slaff) [inline samples/Basic_NFC/app/application.cpp]: The comments below are not relevant and not needed. Please remove them and format (arrange) the code.

File: samples/Basic_NFC/app/application.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
+#include <user_config.h>
+#include <SmingCore/SmingCore.h>
+#include "SerialReadingDelegateDemo.h"
```
Comment (slaff) [inline samples/Basic_NFC/app/application.cpp]: Where are the functions/classes from that file used?

File: samples/Basic_NFC/app/application.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+#include <AppSettings.h>
+#include "MFRC522.h"
+Timer procTimer;
+static Timer nfcScanTimer;
+SerialReadingDelegateDemo delegateDemoClass;
```
Comment (slaff) [inline samples/Basic_NFC/app/application.cpp]: Where is the delegateDemoClass used?

File: samples/Basic_NFC/app/application.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+int helloCounter = 0;
+
+void ICACHE_FLASH_ATTR scanNfc(byte scanner);
+//#define RST_PIN	15 // RST-PIN f�r RC522 - RFID - SPI - Modul GPIO15
+#define SS_PIN	4  	// D2
+//#define SS_PIN	15  // D8
+
+//MFRC522 mfrc522(SS_PIN, RST_PIN);	// Create MFRC522 instance
```
Comment (slaff) [inline samples/Basic_NFC/app/application.cpp]: Please, remove commented out code.

File: samples/Basic_NFC/app/application.cpp [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+void dump_byte_array(byte *buffer, byte bufferSize) {
+  for (byte i = 0; i < bufferSize; i++) {
+    Serial.print(buffer[i] < 0x10 ? " 0" : " ");
+    Serial.print(buffer[i], HEX);
+  }
+}
+//---------------------------------
+void ICACHE_FLASH_ATTR
+scanNfc(byte scanner){
+static int cnt;
```
Comment (slaff) [inline samples/Basic_NFC/app/application.cpp]: Coding style: Please, align that line.

File: samples/Basic_NFC/app/application.cpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	if ( ! mfrc522.PICC_IsNewCardPresent()) {
+		Serial.printf("Scanning nfc Scanner:%d \r\n", scanner);
```
Comment (slaff) [inline samples/Basic_NFC/app/application.cpp]: For consistency it will be better to use only debugf(...) for displaying debug information. Can you, please, replace `Serial.printf()` calls used to display debug information with `debugf()`?

File: samples/Basic_NFC/app/application.cpp [hunk 8]
```diff
@@ -0,0 +1,90 @@
+#include <user_config.h>
+#include <SmingCore/SmingCore.h>
+#include "SerialReadingDelegateDemo.h"
+#include <AppSettings.h>
```
Comment (slaff) [inline samples/Basic_NFC/app/application.cpp]: Why is AppSettings.h included here? Is it used somewhere?

File: samples/Basic_NFC/app/MFRC522.cpp [hunk 9]
```diff
@@ -0,0 +1,1768 @@
+/*
+* MFRC522.cpp - Library to use ARDUINO RFID MODULE KIT 13.56 MHZ WITH TAGS SPI W AND R BY COOQROBOT.
+* NOTE: Please also check the comments in MFRC522.h - they provide useful hints and background information.
+* Released into the public domain.
+*/
+#include <user_config.h>
+#include <SmingCore/SmingCore.h>
+#include <AppSettings.h>
```
Comment (slaff) [inline samples/Basic_NFC/app/MFRC522.cpp]: Why is AppSettings.h included here?

File: samples/Basic_NFC/app/MFRC522.cpp [hunk 10]
```diff
@@ -0,0 +1,1768 @@
+/*
```
Comment (slaff) [inline samples/Basic_NFC/app/MFRC522.cpp]: A better location for MFRC522.cpp and MFRC522.h will be under `Sming/Libraries/MFRC522`. This way also other applications can use those libraries just by including Sming.

File: samples/Basic_NFC/Makefile-user.mk (hunk 1 above)
Comment (icodk) [inline samples/Basic_NFC/Makefile-user.mk]: Will do that
Thanks

File: samples/Basic_NFC/app/application.cpp (hunk 3 above)
Comment (icodk) [inline samples/Basic_NFC/app/application.cpp]: I have no idea
Will check if can be removed

File: samples/Basic_NFC/app/application.cpp (hunk 8 above)
Comment (icodk) [inline samples/Basic_NFC/app/application.cpp]: Removed

File: samples/Basic_NFC/app/application.cpp (hunk 4 above)
Comment (icodk) [inline samples/Basic_NFC/app/application.cpp]: Removed

File: samples/Basic_NFC/app/application.cpp (hunk 5 above)
Comment (icodk) [inline samples/Basic_NFC/app/application.cpp]: Willdo

File: samples/Basic_NFC/app/application.cpp (hunk 6 above)
Comment (icodk) [inline samples/Basic_NFC/app/application.cpp]: Hope I understand what you mean
Restiling my code

File: samples/Basic_NFC/app/application.cpp (hunk 7 above)
Comment (icodk) [inline samples/Basic_NFC/app/application.cpp]: Will do

File: samples/Basic_NFC/app/application.cpp (hunk 2 above)
Comment (icodk) [inline samples/Basic_NFC/app/application.cpp]: Will do

File: samples/Basic_NFC/app/MFRC522.cpp (hunk 10 above)
Comment (icodk) [inline samples/Basic_NFC/app/MFRC522.cpp]: You are probably right but I can't do it right now

File: samples/Basic_NFC/app/MFRC522.cpp (hunk 9 above)
Comment (icodk) [inline samples/Basic_NFC/app/MFRC522.cpp]: Removed
Thanks for your help
Let me know if need more adjustments
I build and run the code and it still works
I will commit it now

File: samples/Basic_NFC/Makefile-user.mk [hunk 11]
```diff
@@ -0,0 +1,40 @@
+## Local build configuration
```
Comment (slaff) [inline samples/Basic_NFC/Makefile-user.mk]: Please, replace the Makefile-user.mk with that one `https://github.com/SmingHub/Sming/blob/develop/samples/Basic_Blink/Makefile-user.mk`, Don't put your local version of Basic_Blink here :)

File: samples/Basic_NFC/include/AppSettings.h [hunk 12]
```diff
@@ -0,0 +1,96 @@
+/*
```
Comment (slaff) [inline samples/Basic_NFC/include/AppSettings.h]: Is that file needed anywhere? If not, please, remove it too.

File: samples/Basic_NFC/include/SerialReadingDelegateDemo.h [hunk 13]
```diff
@@ -0,0 +1,50 @@
+#ifndef INCLUDE_SERIALREADINGDELEGATEDEMO_H_
```
Comment (slaff) [inline samples/Basic_NFC/include/SerialReadingDelegateDemo.h]: Is that file needed anywhere? If not, please, remove it too.

**Comment by icodk** [comment]
@slaff Do you have more requests?
Let me know

**Comment by slaff** [comment]
>  Do you have more requests?

@icodk Only the ones that are still not taken care of:

- A better location for MFRC522.cpp and MFRC522.h will be under Sming/Libraries/MFRC522. This way also other applications can use those libraries just by including Sming
- Remove non-relevant comments in application.cpp (and commented code) and format the code (martk the code and press CTRL+F in Eclipse)

**Comment by icodk** [comment]
Should be fixed now
Let me know

File: samples/Basic_NFC/app/application.cpp [hunk 14]
```diff
@@ -0,0 +1,90 @@
+#include <user_config.h>
+#include <SmingCore/SmingCore.h>
+#include "MFRC522.h"
```
Comment (slaff) [inline samples/Basic_NFC/app/application.cpp]: @icodk May be you should change also the include to point to the right header file
```
#include <Libraries/MFRC522/MFRC522.h>
```

File: samples/Basic_NFC/app/application.cpp (hunk 2 above)
Comment (slaff) [inline samples/Basic_NFC/app/application.cpp]: Don't forget to remove the non-relevant comments on those lines.  If you are wondering which lines then look at that URL (https://github.com/SmingHub/Sming/pull/1040/files) and search for by comments.

File: Sming/Libraries/MFRC522/MFRC522.cpp [hunk 15]
```diff
@@ -0,0 +1,1767 @@
+/*
+* MFRC522.cpp - Library to use ARDUINO RFID MODULE KIT 13.56 MHZ WITH TAGS SPI W AND R BY COOQROBOT.
+* NOTE: Please also check the comments in MFRC522.h - they provide useful hints and background information.
+* Released into the public domain.
+*/
+#include <user_config.h>
+#include <SmingCore/SmingCore.h>
```
Comment (slaff) [inline Sming/Libraries/MFRC522/MFRC522.cpp]: Last, but not least, you should make the code to compile. I would recommend you to move these
includes 
```
#include <user_config.h>
#include <SmingCore/SmingCore.h>
```

from here and put them into the header file `MFRC522.h` as 
```
#include <user_config.h>
#include <SmingCore.h>
```

Commit the changes and make sure that all tests pass.

**Comment by slaff** [comment]
@icodk Your changes will be merged with that rebase #1049. Thanks for your time and efforts.
